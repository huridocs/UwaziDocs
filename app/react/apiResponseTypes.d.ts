import { UserGroupSchema } from 'shared/types/userGroupType';
import { UserSchema } from 'shared/types/userType';
import {
  PreserveConfig,
  Settings,
  SettingsFilterSchema,
  SettingsLinkSchema,
} from 'shared/types/settingsType';
import { LanguageSchema } from 'shared/types/commonTypes';

export interface GroupMemberSchema {
  refId: string;
  username: string;
}

export interface ClientUserGroupSchema extends Omit<UserGroupSchema, '_id' | 'members'> {
  _id?: string;
  members: GroupMemberSchema[];
}

export interface ClientUserSchema extends Omit<UserSchema, '_id' | 'groups'> {
  _id?: string;
  groups?: {
    _id: string;
    name: string;
  }[];
}

export interface ClientSettingsFilterSchema extends Omit<SettingsFilterSchema, '_id'> {
  _id?: string;
}

export interface ClientSublink extends Omit<ClientSettingsLinkSchema, 'sublinks'> {}

export interface ClientSettingsLinkSchema extends Omit<SettingsLinkSchema, '_id' | 'sublinks'> {
  _id?: string;
  sublinks?: ClientSublink[];
}

export interface ClientPreserveConfig extends Omit<PreserveConfig, 'config'> {
  config: {
    token: string;
    template: string;
    user?: string;
  }[];
}

export interface ClientLanguageSchema extends Omit<LanguageSchema, '_id'> {
  _id?: string;
}

export interface ClientSettings
  extends Omit<Settings, '_id | filters | links | features | languages'> {
  _id?: string;
  filters?: ClientSettingsFilterSchema[];
  languages?: ClientLanguageSchema[];
  links?: ClientSettingsLinkSchema[];
  features?: Omit<Settings['features'], 'preserve | metadataExtraction'> & {
    preserve?: ClientPreserveConfig;
    metadataExtraction?: {
      url: string;
      templates?: {
        template: string;
        properties: string[];
      }[];
    };
    [k: string]: unknown | undefined;
  };
}
