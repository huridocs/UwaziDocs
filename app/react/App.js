import React from 'react';
import { Provider } from 'react-redux';
import { routes } from './Routes';
import CustomProvider from './App/Provider';
import { store } from './store';

const App = () => (
  <Provider store={store}>
    <CustomProvider>{routes}</CustomProvider>
  </Provider>
);

export default App;
