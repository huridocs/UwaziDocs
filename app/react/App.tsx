import React from 'react';
import { RouterProvider, createBrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { getRoutes } from './Routes';
import CustomProvider from './App/Provider';
import { store } from './store';

const reduxState = store?.getState();

const router = createBrowserRouter(
  getRoutes(reduxState?.settings.collection.toJS(), reduxState?.user.get('_id'))
);

const App = () => (
  <Provider store={store as any}>
    <CustomProvider>
      <RouterProvider router={router} fallbackElement={null} />
    </CustomProvider>
  </Provider>
);

export { App };
