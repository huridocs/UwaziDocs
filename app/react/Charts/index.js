import loadable from '@loadable/component';

import colorScheme from './utils/colorScheme';
import arrayUtils from './utils/arrayUtils';
import ExtendedTooltip from './components/ExtendedTooltip';

const Bar = loadable(
  async () => import(/* webpackChunkName: "LazyLoadBar" */ './components/Bar.js')
);

const ColoredBar = loadable(
  async () => import(/* webpackChunkName: "LazyLoadColoredBar" */ './components/ColoredBar.js')
);

const Pie = loadable(
  async () => import(/* webpackChunkName: "LazyLoadPie" */ './components/Pie.js')
);

const StackedDualBarChart = loadable(
  async () =>
    import(
      /* webpackChunkName: "LazyLoadStackedDualBarChart" */ './components/StackedDualBarChart.js'
    )
);

export { Bar, ColoredBar, ExtendedTooltip, Pie, StackedDualBarChart, colorScheme, arrayUtils };
