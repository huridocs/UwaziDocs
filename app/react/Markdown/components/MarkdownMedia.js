import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ReactPlayer from 'react-player';

export class MarkdownMedia extends Component {
  propsToConfig(props) {
    const config = {
      url: '',
      options: {}
    };

    let parsedProps = props.config.replace(/\(|\)/g, '').split(',');
    config.url = parsedProps.shift();

    parsedProps = parsedProps.join(',') || '{}';
    parsedProps = parsedProps.replace(/&quot;/g, '"');

    try {
      parsedProps = JSON.parse(parsedProps);
    } catch (error) {
      parsedProps = {};
      console.log(error);
    }

    config.options = parsedProps;

    return config;
  }

  timeLinks(timelinks) {
    timelinks = timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const seconds = timeKey.split(':').reverse().reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * (60 ** _index), 0);
      return (<div className="timelink" key={index} onClick={this.seekTo.bind(this, seconds)} >
        <b><i className="fa fa-play" /> {timeKey}</b>
        <span>{timelinks[timeKey]}</span>
      </div>);
    });
  }

  seekTo(seconds) {
    this.refs.player.seekTo(seconds);
  }

  render() {
    const config = this.propsToConfig(this.props);

    return (
      <div className="video-container">
        <div>
          <ReactPlayer
            ref="player"
            width="100%"
            height="100%"
            url={config.url}
            controls
          />
        </div>
        <div>
          {this.timeLinks(config.options.timelinks)}
        </div>
      </div>
    );
  }
}

MarkdownMedia.propTypes = {
  config: PropTypes.string
};

export default MarkdownMedia;
