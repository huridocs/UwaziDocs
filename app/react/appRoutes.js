import { getRoutes } from './Routes';
import { settingsAtom, atomStore, userAtom } from './V2/atoms';

export const routes = getRoutes && getRoutes(atomStore.get(settingsAtom), atomStore.get(userAtom)?._id);

