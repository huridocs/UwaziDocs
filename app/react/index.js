import React from 'react';
import ReactDOM from 'react-dom';

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

import { AppContainer } from 'react-hot-loader';

import { SENTRY_APP_DSN } from 'app/config';
import App from './App';

import './App/sockets';

if (SENTRY_APP_DSN) {
  Sentry.init({
    dsn: SENTRY_APP_DSN,
    integrations: [new BrowserTracing()],
    tracesSampleRate: 0.1,
  });
}

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
