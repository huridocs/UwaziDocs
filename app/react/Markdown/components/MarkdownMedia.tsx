/* eslint-disable max-lines */
/* eslint-disable max-statements */
import { Translate } from 'app/I18N';
import React, { createRef, LegacyRef, useState } from 'react';
import { FieldArrayWithId, useFieldArray, useForm } from 'react-hook-form';
import ReactPlayer from 'react-player';
import { Icon } from 'UI';

export interface MarkdownMediaProps {
  compact?: boolean;
  editing?: boolean;
  onTimeLinkAdded?: Function;
  config: string;
}

interface TimeLink {
  timeHours: string;
  timeMinutes: string;
  timeSeconds: string;
  label: string;
}

const propsToConfig = (props: MarkdownMediaProps) => {
  const config = { url: '', options: { timelinks: {} } };

  let parsedProps: any = props.config.replace(/\(|\)/g, '').split(',');
  config.url = parsedProps.shift() as string;

  parsedProps = (parsedProps.join(',') || '{}').replace(/&quot;/g, '"');

  try {
    parsedProps = JSON.parse(parsedProps);
  } catch (error) {
    parsedProps = {};
  }

  config.options = parsedProps;

  return config;
};

const formatTimeLinks = (timelinks: any): TimeLink[] =>
  Object.keys(timelinks).map(key => {
    const timeLink = { timeHours: '00', timeMinutes: '00', timeSeconds: '00', label: '' };
    const splitTimes = key.split(':');
    if (splitTimes.length === 2) {
      [timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    } else {
      [timeLink.timeHours, timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    }
    timeLink.label = timelinks[key] as string;
    return timeLink;
  });

const MarkdownMedia = (props: MarkdownMediaProps) => {
  const [newTimeline, setNewTimeline] = useState<TimeLink>({
    timeHours: '00',
    timeMinutes: '00',
    timeSeconds: '00',
    label: '',
  });
  const { options } = propsToConfig(props);
  const originalTimelinks = formatTimeLinks(options?.timelinks || {});
  const [playingTimelinkIndex, setPlayingTimelinkIndex] = useState<number>(-1);
  const [isVideoPlaying, setVideoPlaying] = useState<boolean>(false);

  const { control, register, getValues } = useForm<{ timelines: TimeLink[] }>({
    defaultValues: { timelines: originalTimelinks },
  });
  const { fields, remove, append } = useFieldArray<{ timelines: TimeLink[] }>({
    control,
    name: 'timelines',
  });

  const playerRef: LegacyRef<ReactPlayer> | undefined = createRef();

  const seekTo = (seconds: number) => {
    const playingStatus = isVideoPlaying;
    playerRef.current?.seekTo(seconds);
    setVideoPlaying(playingStatus);
  };

  const constructTimelinksString = (timelinks: TimeLink[], url: string) => {
    const timelinksObj = timelinks.reduce((current: any, timelink) => {
      current[`${timelink.timeHours}:${timelink.timeMinutes}:${timelink.timeSeconds}`] =
        timelink.label;
      return current;
    }, {});
    return `(${url}, ${JSON.stringify({ timelinks: timelinksObj })})`;
  };

  const timeLinks = (_timelinks: any) => {
    const timelinks = _timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const seconds = timeKey
        .split(':')
        .reverse()
        .reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * 60 ** _index, 0);

      let displayTime = timeKey;
      if (displayTime.split(':').length === 2) {
        displayTime = `00:${displayTime}`;
      }

      return (
        <div className="timelink" key={timeKey + index.toString()} title={timelinks[timeKey]}>
          <b
            className="timelink-icon"
            onClick={() => {
              seekTo(seconds);
              if (index === playingTimelinkIndex) {
                setVideoPlaying(false);
                setPlayingTimelinkIndex(-1);
              } else {
                setVideoPlaying(true);
                setPlayingTimelinkIndex(index);
              }
            }}
          >
            <Icon icon={index === playingTimelinkIndex ? 'pause' : 'play'} />
          </b>
          <div
            className="timelink-time-label"
            onClick={() => {
              seekTo(seconds);
              setPlayingTimelinkIndex(index);
            }}
          >
            <b>{displayTime}</b>
            <span>{timelinks[timeKey]}</span>
          </div>
        </div>
      );
    });
  };

  const updateParentForm = (url: string) => {
    const fullTimelinksString = constructTimelinksString(getValues().timelines, url);
    if (props.onTimeLinkAdded) props.onTimeLinkAdded(fullTimelinksString);
  };

  const appendTimelinkAndUpdateParent = (url: string, timelink?: TimeLink) => {
    const currentTimelink = timelink || newTimeline;
    append(currentTimelink);
    updateParentForm(url);
  };

  const renderNewTimeLinkForm = (url: string) => (
    <div className="new-timelink">
      <div className="timestamp">
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeHours: event.target.value });
          }}
          className="timestamp-hours"
          placeholder="00"
          required
          value={newTimeline.timeHours}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeMinutes: event.target.value });
          }}
          className="timestamp-minutes"
          placeholder="00"
          required
          value={newTimeline.timeMinutes}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeSeconds: event.target.value });
          }}
          className="timestamp-seconds"
          placeholder="00"
          required
          value={newTimeline.timeSeconds}
        />
      </div>
      <input
        type="text"
        onChange={event => {
          setNewTimeline({ ...newTimeline, label: event.target.value });
        }}
        className="timestamp-label"
        placeholder="Enter title"
        value={newTimeline.label}
      />
      <button
        type="button"
        className="new-timestamp-btn"
        onClick={() => {
          appendTimelinkAndUpdateParent(url);
          setNewTimeline({ timeHours: '00', timeMinutes: '00', timeSeconds: '00', label: '' });
        }}
      >
        <Icon icon="plus" />
      </button>
    </div>
  );

  const renderSingleTimeLinkInputs = (
    field: FieldArrayWithId<TimeLink>,
    url: string,
    index: number
  ) => (
    <div className="new-timelink" key={index}>
      <div className="timestamp">
        <input
          type="text"
          className="timestamp-hours"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeHours`, {
            onChange: _ => updateParentForm(url),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-minutes"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeMinutes`, {
            onChange: _ => updateParentForm(url),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-seconds"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeSeconds`, {
            onChange: _ => updateParentForm(url),
          })}
        />
      </div>
      <input
        type="text"
        className="timestamp-label"
        placeholder="Enter title"
        key={field.id}
        {...register(`timelines.${index}.label`, {
          onChange: _ => updateParentForm(url),
        })}
      />
      <button
        title="Remove timelink"
        type="button"
        className="delete-timestamp-btn"
        onClick={() => {
          remove(index);
          updateParentForm(url);
        }}
      >
        <Icon icon="trash-alt" />
      </button>
    </div>
  );

  const renderTimeLinksForm = (url: string) => (
    <>
      {fields.map((field: FieldArrayWithId<TimeLink>, index: number) =>
        renderSingleTimeLinkInputs(field, url, index)
      )}
      {renderNewTimeLinkForm(url)}
    </>
  );

  const config = propsToConfig(props);
  const { compact, editing } = props;
  const dimensions: { width: string; height?: string } = { width: '100%' };
  if (compact) {
    dimensions.height = '100%';
  }

  return (
    <div className={`video-container ${compact ? 'compact' : ''}`}>
      <div>
        <ReactPlayer
          className="react-player"
          playing={isVideoPlaying}
          ref={playerRef}
          url={config.url}
          {...dimensions}
          controls
          onPause={() => {
            setVideoPlaying(false);
          }}
          onPlay={() => {
            setVideoPlaying(true);
          }}
        />
      </div>
      {!editing && <div>{timeLinks(config.options.timelinks)}</div>}
      {editing && (
        <div className="timelinks-form">
          <button
            type="button"
            className="add-timelink"
            onClick={() => {
              const currentTime = playerRef.current?.getCurrentTime() as number;
              const hours = Math.floor(currentTime / 3600);
              const remainingSeconds = currentTime - hours * 3600;
              const minutes = Math.floor(remainingSeconds / 60);
              const seconds = Math.floor(remainingSeconds % 60);
              const timelink = {
                timeHours: hours < 10 ? `0${hours.toString()}` : hours.toString(),
                timeMinutes: minutes < 10 ? `0${minutes.toString()}` : minutes.toString(),
                timeSeconds: seconds < 10 ? `0${seconds.toString()}` : seconds.toString(),
                label: '',
              };
              appendTimelinkAndUpdateParent(config.url, timelink);
            }}
          >
            <Translate>Add timelink</Translate>
          </button>
          <h5>
            <Translate>Timelinks</Translate>
          </h5>
          {renderTimeLinksForm(config.url)}
        </div>
      )}
      <p className="print-view-alt">{config.url}</p>
    </div>
  );
};

export default MarkdownMedia;
