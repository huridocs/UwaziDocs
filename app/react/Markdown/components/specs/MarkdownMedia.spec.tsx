/**
 * @jest-environment jsdom
 */
import React from 'react';
import { RenderResult, fireEvent, screen, act } from '@testing-library/react';
import fetchMock from 'fetch-mock';

import { defaultState, renderConnectedContainer } from 'app/utils/test/renderConnected';
import MarkdownMedia from '../MarkdownMedia';

let playerRef: { current: { seekTo: any; getCurrentTime: () => number } };

const mockUseRef = jest.fn().mockImplementation(args => {
  const actual = jest.requireActual<typeof React>('react')
    ? jest.requireActual<typeof React>('react').useRef(args)
    : null;
  if (args === null && actual) {
    playerRef = actual;
    return actual;
  }
  return actual;
});

jest.mock('react', () => ({
  ...jest.requireActual<typeof React>('react'),
  useRef: (args: any) => mockUseRef(args),
}));

describe('MarkdownMedia', () => {
  let renderResult: RenderResult;
  const mockedCreateObjectURL: jest.Mock = jest.fn();
  const mockedRevokeObjectURL: jest.Mock = jest.fn();

  beforeAll(() => {
    URL.createObjectURL = mockedCreateObjectURL;
    mockedCreateObjectURL.mockReturnValue('blob:abc');
    URL.revokeObjectURL = mockedRevokeObjectURL;
    fetchMock.mock('https://www.vimeo.com/253530307', {}, { overwriteRoutes: true });
  });

  afterAll(() => {
    mockedCreateObjectURL.mockReset();
    mockedRevokeObjectURL.mockReset();
  });
  const onTimeLinkAdded = jest.fn();
  const render = async (
    options: { compact?: boolean; editing?: boolean } = { compact: false, editing: false }
  ) => {
    const props = {
      ...options,
      onTimeLinkAdded,
      config:
        '(https://www.vimeo.com/253530307, {"timelinks": {"02:10": "A rude awakening", "05:30": "Finally, you are up!"}})',
    };
    await act(() => {
      ({ renderResult } = renderConnectedContainer(
        <MarkdownMedia {...props} />,
        () => defaultState
      ));
    });
  };

  describe('render', () => {
    it('should render an iframe with the correct video id', async () => {
      await render();
      expect(renderResult.asFragment()).toMatchSnapshot();
      expect(mockedCreateObjectURL.mock.calls[0].toString()).toEqual('[object Blob]');
    });
    it('should revoke the created URL', async () => {
      await render();
      renderResult.unmount();
      expect(mockedRevokeObjectURL).toHaveBeenCalledWith('blob:abc');
    });
    it('should render the edition mode', async () => {
      await render({ editing: true });
      expect(renderResult.asFragment()).toMatchSnapshot();
    });
    it('should use compact class name if compact prop is set', async () => {
      await render({ compact: true });
      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    describe('Video timeline', () => {
      it('should render the timelinks', async () => {
        await render();
        const links = renderResult.container.getElementsByClassName('timelink');
        expect(links.length).toBe(2);
      });

      it('should interact with the player through time line label', async () => {
        await render();
        const spySeekTo = jest.spyOn(playerRef.current, 'seekTo');
        const firstTimeLink = screen.getByText('A rude awakening').parentElement!;

        await act(async () => {
          fireEvent.click(firstTimeLink);
          expect(spySeekTo).toHaveBeenCalledWith(2 * 60 + 10);
        });
        const secondTimeLink = screen.getByText('Finally, you are up!').parentElement!;

        await act(async () => {
          fireEvent.click(secondTimeLink);
          expect(spySeekTo).toHaveBeenCalledWith(5 * 60 + 30);
        });
      });

      const playTimeLink = async (
        index: number,
        duration: number,
        status: string,
        spySeekTo: jest.SpyInstance
      ) => {
        const timeLink = renderResult.container.getElementsByClassName('timelink-icon')[index];
        await act(async () => {
          fireEvent.click(timeLink);
          expect(spySeekTo).toHaveBeenCalledWith(duration);
          const icon = renderResult.container.getElementsByClassName('timelink-icon')[index];
          expect(icon.children[0].getAttribute('data-icon')).toEqual(status);
        });
      };
      it('should interact with the player through time line icon', async () => {
        await render();
        const spySeekTo = jest.spyOn(playerRef.current, 'seekTo');
        await playTimeLink(0, 2 * 60 + 10, 'play', spySeekTo);
        await playTimeLink(1, 5 * 60 + 30, 'play', spySeekTo);
        await playTimeLink(1, 5 * 60 + 30, 'pause', spySeekTo);
      });
      const editTimeLink = async (inputs: HTMLElement[], value: string, field: string) => {
        await act(async () => {
          fireEvent.change(inputs.find(x => (x as HTMLInputElement).name === field)!, {
            target: { value },
          });
        });
      };

      it('should allow to add a timelink', async () => {
        await render({ editing: true });
        jest.spyOn(playerRef.current, 'getCurrentTime').mockReturnValue(27784);
        const addLinkBtn = screen.getByText('Add timelink').parentElement!;
        await act(async () => {
          fireEvent.click(addLinkBtn);
        });
        await act(async () => {
          await editTimeLink(
            screen.getAllByRole('textbox'),
            'Check at this!',
            'timelines.2.label'
          )!;
          expect(onTimeLinkAdded).toHaveBeenCalledWith([
            { label: 'A rude awakening', timeHours: '00', timeMinutes: '02', timeSeconds: '10' },
            {
              label: 'Finally, you are up!',
              timeHours: '00',
              timeMinutes: '05',
              timeSeconds: '30',
            },
            { label: 'Check at this!', timeHours: '07', timeMinutes: '43', timeSeconds: '04' },
          ]);
        });
      });

      it('should allow to edit a saved timelink', async () => {
        await render({ editing: true });
        const removeTimeLinkButtons = screen.getAllByRole('button', { name: 'Remove timelink' });

        await act(async () => {
          fireEvent.click(removeTimeLinkButtons[0]);
        });
        await act(async () => {
          const inputs = screen.getAllByRole('textbox');
          await editTimeLink(inputs, '04', 'timelines.0.timeHours');
          await editTimeLink(inputs, '42', 'timelines.0.timeMinutes');
          await editTimeLink(inputs, '56', 'timelines.0.timeSeconds');
        });
        await act(async () => {
          expect(onTimeLinkAdded).toHaveBeenCalledWith([
            {
              label: 'Finally, you are up!',
              timeHours: '04',
              timeMinutes: '42',
              timeSeconds: '56',
            },
          ]);
        });
      });

      it('should allow to edit a new timelink', async () => {
        await render({ editing: true });
        const addLinkBtn = screen.getByText('Add timelink').parentElement!;
        await act(async () => {
          fireEvent.click(addLinkBtn);
        });
        const updatedInputs = screen.getAllByRole('textbox');
        await editTimeLink(updatedInputs, '12', 'timelines.2.timeHours');
        await editTimeLink(updatedInputs, '17', 'timelines.2.timeMinutes');
        await editTimeLink(updatedInputs, '34', 'timelines.2.timeSeconds');
        await editTimeLink(updatedInputs, 'new check point', 'timelines.2.label');
        expect(onTimeLinkAdded).toHaveBeenCalledWith([
          { label: 'A rude awakening', timeHours: '00', timeMinutes: '02', timeSeconds: '10' },
          {
            label: 'Finally, you are up!',
            timeHours: '00',
            timeMinutes: '05',
            timeSeconds: '30',
          },
          { label: 'new check point', timeHours: '12', timeMinutes: '17', timeSeconds: '34' },
        ]);
      });
    });
  });
});
