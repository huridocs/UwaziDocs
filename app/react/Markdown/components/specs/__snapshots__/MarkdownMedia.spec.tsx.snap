// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownMedia render should render an iframe with the correct video id 1`] = `
<DocumentFragment>
  <div
    class="video-container "
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 360px;"
      />
    </div>
    <div>
      <div
        class="timelink"
        title="A rude awakening"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:02:10
          </b>
          <span>
            A rude awakening
          </span>
        </div>
      </div>
      <div
        class="timelink"
        title="Finally, you are up!"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:05:30
          </b>
          <span>
            Finally, you are up!
          </span>
        </div>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      https://www.vimeo.com/253530307
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkdownMedia render should use compact class name if compact prop is set 1`] = `
<DocumentFragment>
  <div
    class="video-container compact"
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 100%;"
      >
        <div
          style="width: 100%; height: 100%; overflow: hidden;"
        />
      </div>
    </div>
    <div>
      <div
        class="timelink"
        title="A rude awakening"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:02:10
          </b>
          <span>
            A rude awakening
          </span>
        </div>
      </div>
      <div
        class="timelink"
        title="Finally, you are up!"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:05:30
          </b>
          <span>
            Finally, you are up!
          </span>
        </div>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      https://www.vimeo.com/253530307
    </p>
  </div>
</DocumentFragment>
`;
